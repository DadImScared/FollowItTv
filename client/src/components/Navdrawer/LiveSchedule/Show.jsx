
import React from 'react';
import PropTypes from 'prop-types';

import moment from 'moment';

import { ListItem, ListItemText } from 'material-ui/List';

import Countdown from '../../Countdown';


const Show = ({ show, showTimer, moveShow, currentStep, eventTime, episodeId, ...other }) => (
  <ListItem {...other}>
    <ListItemText
      primary={show.name}
      secondary={
        showTimer ?
          <Countdown
            callBack={() => moveShow(currentStep, episodeId)}
            eventTime={
              currentStep === 'currentlyAiring' ?
                moment(eventTime, 'YYYY-MM-DD HH:mm').add(show.runtime, 'minutes')
                :
                moment(eventTime, 'HH:mm')
            }
          />
          :
          eventTime
      }
    />
  </ListItem>
);

Show.propTypes = {
  show: PropTypes.object.isRequired,
  moveShow: PropTypes.func.isRequired,
  eventTime: PropTypes.string.isRequired,
  episodeId: PropTypes.number.isRequired,
  showTimer: PropTypes.bool,
  currentStep: PropTypes.string
};

Show.defaultProps = {
  showTimer: true,
  currentStep: 'willAir'
};

export default Show;
