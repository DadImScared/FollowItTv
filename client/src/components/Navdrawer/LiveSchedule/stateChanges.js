import _ from 'lodash';
import moment from 'moment/moment';

export const organizeTodayShows = (state, props) => {
  const willAir = new Set();
  const hasAired = new Set();
  const isAiring = [];
  const { schedule , episodes } = props;
  const { today } = state;
  const now = moment();
  schedule[today].forEach((episodeId) => {
    const { airtime: time, airdate: date, runtime } = episodes[episodeId];
    const showDate = moment(`${date} ${time}`, 'YYYY-MM-DD HH:mm');
    if (now.diff(showDate) < 0) {
      willAir.add(episodeId);
    }
    else if (now.diff(showDate.clone().add(runtime, 'minutes')) > 0 ) {
      hasAired.add(episodeId);
    }
    else {
      isAiring.push(episodeId);
    }
  });
  return {
    willAir: [...willAir],
    currentlyAiring: _.union(state.currentlyAiring, isAiring),
    hasAired: [...hasAired]
  };
};

export const getYesterdayIsAiring = (state, props) => {
  const { episodes, schedule } = props;
  const now = moment();
  const yesterday = now.clone().subtract(1, 'days').format('YYYY-MM-DD');
  const yesterdayisAiring = [];
  schedule[yesterday].forEach((episodeId) => {
    const { airtime: time, airdate: date, runtime } = episodes[episodeId];
    const showDate = moment(`${date} ${time}`, 'YYYY-MM-DD HH:mm');
    if (now.diff(showDate) < 0 || now.diff(showDate.clone().add(runtime, 'minutes')) > 0) {
      return;
    }
    yesterdayisAiring.push(episodeId);
  });
  return {
    currentlyAiring: _.union([...state.currentlyAiring, ...yesterdayisAiring])
  };
};
