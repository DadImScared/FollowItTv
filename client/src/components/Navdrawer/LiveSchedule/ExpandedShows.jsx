
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { withStyles } from 'material-ui/styles';
import ExpansionPanel, {
  ExpansionPanelDetails,
  ExpansionPanelSummary
} from 'material-ui/ExpansionPanel';
import List from 'material-ui/List';
import Paper from 'material-ui/Paper';
import Typography from 'material-ui/Typography';

import ExpandMoreIcon from 'material-ui-icons/ExpandMore';

import Show from './Show';
import { LiveSchedule } from '../../../styles/Navdrawer';

const { expandedShows: styles } = LiveSchedule;


export class ExpandedShows extends Component {
  constructor(...args) {
    super(...args);
    this.state = {
      expanded: false
    };
  }

  componentDidMount() {
    const { panels } = this.props;
    this.changePanel(true, panels[0][0].length ? 0:1);
  }

  changePanel = (expanded, panel) => {
    this.setState({
      expanded: expanded ? panel:false
    });
  };

  render() {
    const { panels, shows, episodes, moveShow, classes } = this.props;
    const { expanded } = this.state;
    return (
      <div>
        {
          panels.map(([panel, description], index) => {
            const panelName = description === 'Airing soon' ? 'willAir':'hasAired';
            return (
              <ExpansionPanel
                expanded={expanded === index}
                onChange={(event, expanded) => this.changePanel(expanded, index)}
                key={`${description}-${index}`}>
                <ExpansionPanelSummary classes={{ root: classes[`${panelName}Summary`] }} expandIcon={<ExpandMoreIcon />}>
                  <Typography>
                    {description}
                  </Typography>
                </ExpansionPanelSummary>
                <ExpansionPanelDetails classes={{ root: classes[`${panelName}Detail`] }}>
                  <Paper className={classes[`${panelName}List`]}>
                    <List className={classes.innerList}>
                      {
                        episodes ?
                          panel.map((item, index) => {
                            const show = shows[episodes[item].show] || {};
                            return (
                              <Show
                                episodeId={item}
                                key={`${item}-${index}`}
                                showTimer={show.schedule && description === 'Airing soon'}
                                show={show}
                                moveShow={moveShow}
                                eventTime={episodes[item].airtime}
                              />
                            );
                          })
                          :
                          <div>loading</div>
                      }
                    </List>
                  </Paper>
                </ExpansionPanelDetails>
              </ExpansionPanel>
            );
          })
        }
      </div>
    );
  }
}

ExpandedShows.propTypes = {
  panels: PropTypes.array.isRequired,
  shows: PropTypes.object.isRequired,
  moveShow: PropTypes.func.isRequired
};

export default withStyles(styles)(ExpandedShows);

