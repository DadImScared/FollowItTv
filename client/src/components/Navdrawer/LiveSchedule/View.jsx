
import React from 'react';

import ExpandedShows from './ExpandedShows';


export const View = ({ today, currentlyAiring, willAir, hasAired, shows }) => (
  <div>
    <div>
      currently airing
    </div>
    {
      willAir.length || hasAired.length ?
        <ExpandedShows
          shows={shows}
          panels={[
            [(willAir.length && willAir) || [], 'Airing soon'],
            [(hasAired.length && hasAired) || [], 'Already aired']
          ]}
        />
        :
        <div>loading</div>
    }
  </div>
);

export default View;
