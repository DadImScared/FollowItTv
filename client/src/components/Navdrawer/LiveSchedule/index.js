
import React, { Component } from 'react';

import _ from 'lodash';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import moment from 'moment';
import PropTypes from 'prop-types';

import { addShows } from '../../../actions/shows';
import { getSchedule, getEpisodesAndShows, addSchedule } from '../../../actions/schedule';
import { addEpisodes } from '../../../actions/episodes';

import View from './View';


const nextStep = {
  'willAir': 'currentlyAiring',
  'currentlyAiring': 'hasAired'
};


export class LiveSchedule extends Component {
  constructor(...args) {
    super(...args);
    this.state = {
      today: '',
      currentlyAiring: [],
      hasAired: [],
      willAir: [],
      errorMessage: ''
    };
  }

  componentDidMount() {
    const { schedule } = this.props;
    const today = this.today();
    this.setState({ today }, async () => {
      if (!schedule[today]) {
        await this.fetchSchedule();
      }
      try {
        this.organizeEpisodes();
      }
      catch (e) {
        console.log(e);
        // for some reasons shows didn't load
        // maybe timeout from api so prompt for refresh
        this.setState({ errorMessage: 'there has been an error loading shows please refresh and try again' });
      }
    });
  }

  organizeEpisodes = () => {
    // organize shows into one of category
    // "willAir" "currentlyAiring" "hasAired"
    const willAir = [];
    const currentlyAiring = [];
    const hasAired = [];
    const { schedule , episodes } = this.props;
    const { today } = this.state;
    const now = moment();
    schedule[today].forEach((episodeId) => {
      const { airtime: time, airdate: date, runtime } = episodes[episodeId];
      const showDate = moment(`${date} ${time}`, 'YYYY-MM-DD HH:mm');
      if (now.diff(showDate) < 0) {
        this.pushArray(willAir, episodeId);
      }
      else if (now.diff(showDate.clone().add(runtime, 'minutes')) > 0 ) {
        this.pushArray(hasAired, episodeId);
      }
      else {
        this.pushArray(currentlyAiring, episodeId);
      }
    });
    this.setState({ willAir, currentlyAiring, hasAired });
  };

  pushArray = (arr, id) => {
    if (!arr.includes(id)) {
      arr.push(id);
    }
  };

  moveShow = (currentStep, id) => {
    this.setState((prevState, { episodes }) => {
      const currentStepList = prevState[currentStep].filter((item) => item !== id);
      const nextStepList = [...prevState[nextStep[currentStep]], id];
      const sortedList = _.orderBy(nextStepList, [(o) => moment(episodes[o].airtime, 'HH:mm')], ['desc']);
      return {
        [currentStep]: currentStepList,
        [nextStep[currentStep]]: sortedList
      };
    });
  };

  fetchSchedule = async () => {
    const { addShows, addEpisodes, addSchedule } = this.props;
    const { today } = this.state;
    try {
      const { data } = await getSchedule(today);
      const { episodes, shows, episodeIds } = getEpisodesAndShows(data);
      addShows(shows);
      addEpisodes(episodes);
      addSchedule(today, episodeIds);
    }
    catch (e) {
      console.log(e);
    }
  };

  today = () => {
    return moment().format('YYYY-MM-DD');
  };

  render() {
    const { episodes, shows } = this.props;
    return (
      <View {...this.state} episodes={episodes} shows={shows} moveShow={this.moveShow} />
    );
  }
}

LiveSchedule.propTypes = {
  schedule: PropTypes.object.isRequired,
  shows: PropTypes.object.isRequired,
  episodes: PropTypes.object.isRequired,
  addShows: PropTypes.func.isRequired,
  addEpisodes: PropTypes.func.isRequired,
  addSchedule: PropTypes.func.isRequired
};

const mapStateToProps = ({ shows, schedule, episodes }) => ({
  schedule,
  shows,
  episodes
});

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({
    addShows,
    addEpisodes,
    addSchedule
  }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(LiveSchedule);
