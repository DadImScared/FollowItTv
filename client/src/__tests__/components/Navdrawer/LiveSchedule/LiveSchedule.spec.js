
import React from 'react';

import axios from 'axios';
import moment from 'moment';
import sinon from 'sinon';
import { shallow } from 'enzyme';

import { LiveSchedule } from '../../../../components/Navdrawer/LiveSchedule/index';

jest.mock('axios');


const data = [];


describe('LiveSchedule', () => {
  let wrapper;
  let instance;
  let props;

  beforeEach(() => {
    axios.get.mockImplementation(() => ({ data }));
    props = {
      schedule: {},
      shows: {},
      episodes: {},
      addShows: sinon.spy(),
      addEpisodes: sinon.spy(),
      addSchedule: sinon.spy()
    };
    wrapper = shallow(<LiveSchedule {...props} />);
    instance = wrapper.instance();
  });

  it('should set todays date on mount', () => {
    const expectedDate = moment().format('YYYY-MM-DD');
    expect(wrapper.state().today).toEqual(expectedDate);
  });

  it('should fetchSchedule if no schedule for today', () => {
    const spiedMethod = sinon.spy(instance, 'fetchSchedule');
    instance.componentDidMount();
    expect(spiedMethod.calledOnce).toEqual(true);
  });

  test('organizeEpisodes method', () => {
    // this test might fail if ran at 12:00am to 12:10am will have to find solution in future
    // a normal schedule will not fail regardless of time
    const today = moment().format('YYYY-MM-DD');
    const episodes = {
      1: {
        id: 1,
        show: 1,
        runtime: 30,
        airtime: moment().add(1, 'minutes').format('HH:mm'),
        airdate: today
      },
      2: {
        id: 2,
        show: 2,
        runtime: 60,
        airtime: moment().subtract(20, 'minutes').format('HH:mm'),
        airdate: today
      },
      3: {
        id: 3,
        show: 3,
        runtime: 10,
        airtime: moment().subtract(15, 'minutes').format('HH:mm'),
        airdate: today
      }
    };
    const shows = {
      1: {
        id: 1,
        name: 'breaking bad',
        schedule: {
          time: moment().add(1, 'minutes').format('HH:mm')
        },
        runtime: 30
      },
      2: {
        id: 2,
        name: 'under the dome',
        schedule: {
          time: moment().subtract(20, 'minutes').format('HH:mm')
        },
        runtime: 60
      },
      3: {
        id: 3,
        name: 'show3',
        schedule: {
          time: moment().subtract(15, 'minutes').format('HH:mm')
        },
        runtime: 10
      }
    };
    props = {
      ...props,
      schedule: {
        [today]: [1, 2, 3]
      },
      episodes,
      shows
    };
    wrapper = shallow(<LiveSchedule {...props} />);
    const state = wrapper.state();
    expect(state.currentlyAiring.includes(2)).toEqual(true);
    expect(state.willAir.includes(1)).toEqual(true);
    expect(state.hasAired.includes(3)).toEqual(true);
  });

  test('moveShows', () => {
    const episodes = {
      1: {
        id: 1,
        name: 'breaking bad',
        airtime: moment().add(1, 'minutes').format('HH:mm')
      },
      2: {
        id: 2,
        name: 'under the dome',
        airtime: moment().subtract(20, 'minutes').format('HH:mm')
      },
      3: {
        id: 3,
        name: 'show3',
        airtime: moment().subtract(15, 'minutes').format('HH:mm')
      }
    };
    wrapper = shallow(<LiveSchedule {...props} episodes={episodes} />);
    instance = wrapper.instance();
    instance.setState({ currentlyAiring: [1], hasAired: [2], willAir: [3] });
    instance.moveShow('currentlyAiring', 1);
    expect(instance.state.currentlyAiring).toEqual([]);
    expect(instance.state.hasAired).toEqual([1, 2]);
    instance.moveShow('willAir', 3);
    expect(instance.state.currentlyAiring).toEqual([3]);
  });

  test('multiple moveShow calls at same time', () => {
    const clock = sinon.useFakeTimers();
    const episodes = {
      1: {
        id: 1,
        name: 'breaking bad',
        airtime: moment().add(1, 'minutes').format('HH:mm')
      },
      2: {
        id: 2,
        name: 'under the dome',
        airtime: moment().subtract(20, 'minutes').format('HH:mm')
      },
      3: {
        id: 3,
        name: 'show3',
        airtime: moment().subtract(15, 'minutes').format('HH:mm')
      }
    };
    wrapper = shallow(<LiveSchedule {...props} episodes={episodes} />);
    instance = wrapper.instance();
    instance.setState({ currentlyAiring: [1, 3], hasAired: [2] });
    setTimeout(() => instance.moveShow('currentlyAiring', 1), 1000);
    setTimeout(() => instance.moveShow('currentlyAiring', 3), 1000);
    clock.tick(1000);
    expect(instance.state.hasAired).toEqual([1, 3, 2]);
    expect(instance.state.currentlyAiring).toEqual([]);
  });
});
