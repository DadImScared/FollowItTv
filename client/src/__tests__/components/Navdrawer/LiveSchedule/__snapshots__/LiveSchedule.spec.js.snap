// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveSchedule getYesterdayisAiring 1`] = `
Object {
  "currentlyAiring": Array [
    1,
  ],
}
`;
